File: app/src/main/java/com/example/newsapp/data/repository/NewsRepositoryImpl.kt
package com.example.newsapp.data.repository

import com.example.newsapp.data.mapper.toDomain
import com.example.newsapp.data.remote.NewsApiService
import com.example.newsapp.domain.model.Article
import com.example.newsapp.domain.repository.NewsRepository

class NewsRepositoryImpl(
    private val apiService: NewsApiService,
    private val apiKey: String
) : NewsRepository {

    private var cachedArticles: List<Article> = emptyList()

    override suspend fun getTopHeadlines(category: String?, query: String?): Result<List<Article>> {
        return try {
            val response = apiService.getTopHeadlines(
                apiKey = apiKey,
                category = category,
                query = query
            )
            val articles = response.articles.orEmpty().map { it.toDomain().copy(category = category ?: it.category) }
            cachedArticles = articles
            Result.success(articles)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override fun searchCachedHeadlines(keyword: String): List<Article> {
        return cachedArticles.filter { article ->
            article.title.contains(keyword, ignoreCase = true) ||
                article.description.contains(keyword, ignoreCase = true)
        }
    }
}

File: lib/data/repository/news_repository_impl.dart
import '../../domain/models/article.dart';
import '../../domain/repository/news_repository.dart';
import '../models/article_dto.dart';
import '../remote/news_api_service.dart';

class NewsRepositoryImpl implements NewsRepository {
  NewsRepositoryImpl(this._service);

  final NewsApiService _service;
  List<Article> _cached = [];

  @override
  Future<List<Article>> getTopHeadlines({String? category}) async {
    final response = await _service.fetchTopHeadlines(category: category);
    _cached = response.articles.map(_mapArticle).toList();
    return _cached;
  }

  @override
  Future<List<Article>> searchHeadlines(String query) async {
    if (query.isEmpty) {
      return _cached;
    }
    try {
      final response = await _service.fetchTopHeadlines(query: query);
      final articles = response.articles.map(_mapArticle).toList();
      _cached = articles;
      return articles;
    } catch (_) {
      return _cached
          .where((article) =>
              article.title.toLowerCase().contains(query.toLowerCase()) ||
              article.description.toLowerCase().contains(query.toLowerCase()))
          .toList();
    }
  }

  @override
  List<Article> cachedHeadlines() => List.unmodifiable(_cached);

  Article _mapArticle(ArticleDto dto) => Article(
        title: dto.title,
        author: dto.author,
        description: dto.description,
        imageUrl: dto.urlToImage,
        source: dto.source,
      );
}
